add useUnique feature for buffering last status data

zxcpoiu opened the pull request:
This feature is for buffering which have to be unique to specific key.
a common use case is when you want to log user's last activity to database in bulk way, or latest status such as presence status used in Instant Message.

this commit also implement a flushQueue to reduce re-ordering overhead when calling undequeue()

zxcpoiu committed:
add useUnique feature for buffering last status data

peecky commented:
Thanks for the new feature.

However, I found that the undequeue() function works differently from the previous version.

Test code:

```
var Buffering = require('node-buffering');
var buffering = new Buffering();
buffering.on('flush', function(data) { console.log(data); });
buffering.enqueue(['job1']);
buffering.enqueue(['job2', 'job3']);
buffering.undequeue(['imbusy']);
buffering.undequeue(['imtoobusy']);
buffering.flush();
```

Result of previous version:

```
[ 'imtoobusy', 'imbusy', 'job1', 'job2', 'job3' ]
```

zxcpoiu committed:
FIX: general enqueue() data is FIFO; and undequeue() is LIFO.
FIX: a flush() call will flush static amount of items specified by sizeThreshold, otherwise flush all.

zxcpoiu commented:
Hi peecky,

I fixed the behavior accordingly.
thanks to all your contributions!
